query-planner-js/src/buildPlan.ts

```diff
@@ -859,6 +883,100 @@ class FetchGroup {
   }
 }
 
+function addTypenameFieldForAbstractTypesInNamedFragments(fragments: NamedFragments): NamedFragments {
+  // This method is a bit tricky due to potentially nested fragments. More precisely, suppose that
+  // we have:
+  //   fragment MyFragment on T {
+  //     a {
+  //       b {
+  //         ...InnerB
+  //       }
+  //     }
+  //   }
+  //
+  //   fragment InnerB on B {
+  //     __typename
+  //     x
+  //     y
+  //   }
+  // then if we were to "naively" add `__typename`, the first fragment would end up being:
+  //   fragment MyFragment on T {
+  //     a {
+  //       __typename
+  //       b {
+  //         __typename
+  //         ...InnerX
+  //       }
+  //     }
+  //   }
+  // but that's not ideal because the inner-most `__typename` is already within `InnerX`. And that 
+  // gets in the way to re-adding fragments (the `SelectionSet.optimize` method) because if we start
+  // with:
+  //   {
+  //     a {
+  //       __typename
+  //       b {
+  //         __typename
+  //         x
+  //         y
+  //       }
+  //     }
+  //   }
+  // and add `InnerB` first, we get:
+  //   {
+  //     a {
+  //       __typename
+  //       b {
+  //         ...InnerB
+  //       }
+  //     }
+  //   }
+  // and it becomes tricky to recognize the "updated-with-typename" version of `MyFragment` now (we "seem"
+  // to miss a `__typename`).
+  //
+  // Anyway, to avoid this issue, what we do is that for every fragment, we:
+  //  1. expand any nested fragments in its selection.
+  //  2. add `__typename` where we should in that expanded selection.
+  //  3. re-optimize all fragments (using the "updated-with-typename" versions).
+  // which ends up getting us what we need. However, doing so requires us to deal with fragments in order
+  // of dependencies (first the ones with no nested fragments, then the one with only nested fragments of
+  // that first group, etc...), and that's why this method is a bit longer that one could have expected.
+  type FragmentInfo = {
+    original: NamedFragmentDefinition,
+    expandedSelectionSet: SelectionSet,
+    dependentsOn: string[],
+  };
+  const fragmentsMap = new Map<string, FragmentInfo>();
+
+  for (const fragment of fragments.definitions()) {
+    const expandedSelectionSet = fragment.selectionSet.expandFragments();
+    addTypenameFieldForAbstractTypes(expandedSelectionSet);
+    const otherFragmentsUsages = new Map<string, number>();
+    fragment.collectUsedFragmentNames(otherFragmentsUsages);
+    fragmentsMap.set(fragment.name, {
+      original: fragment,
+      expandedSelectionSet,
+      dependentsOn: Array.from(otherFragmentsUsages.keys()),
+    });
+  }
+
+  const optimizedFragments = new NamedFragments();
+  while (fragmentsMap.size > 0) {
+    for (const [name, info] of fragmentsMap) {
+      // Note that graphQL specifies that named fragments cannot have cycles (https://spec.graphql.org/draft/#sec-Fragment-spreads-must-not-form-cycles)
```

Comment:
**question** Does this mean that if a user _does_ try to send a query like this that it'll be caught somewhere else?

Reply:
I assume it does :). That is, queries are validated though the graphQL-js validations so unless graphQL-js forgot to validate that one, we're good. I'll admit I didn't double-checked, but I'd be super surprised if it was an oversight of graphQL-js given how bad this could blow up any server.

Reply:
Gotcha. I had to test it to see how it works for myself 😈 `apollo-server-core` calls off to `graphql-js` to do validation and throws an error and we end up with something like this `GraphQLError: Cannot spread fragment \"edibleFrag\" within itself via \"hasStemFrag\"`.

It does look like `parseOperation` called on a fragment with a named cycle will crash with a maximum recursion limit, but that's probably not worth dealing with since we don't expect people to call that directly.